pkg/prometheus/prometheus.go

```diff
@@ -37,37 +28,102 @@ var ResponseSizeBuckets = append(append(append(append(
 	prometheus.LinearBuckets(1000000, 1000000, 5)...),
 )
 
-// WithTelemetry instruments the HTTP server with prometheus
-func WithTelemetry(handler http.Handler) http.HandlerFunc {
-	counter := prometheus.NewCounterVec(
+var (
+	// server metrics
+	counter = prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "http_requests_total",
 			Help: "A counter for requests to the wrapped handler.",
 		},
-		[]string{"code"},
+		[]string{"code", "method"},
 	)
 
-	duration := prometheus.NewHistogramVec(
+	duration = prometheus.NewHistogramVec(
 		prometheus.HistogramOpts{
 			Name:    "http_request_duration_seconds",
 			Help:    "A histogram of latencies for requests in seconds.",
 			Buckets: RequestDurationBucketsSeconds,
 		},
-		[]string{"code"},
+		[]string{"code", "method"},
 	)
 
-	responseSize := prometheus.NewHistogramVec(
+	responseSize = prometheus.NewHistogramVec(
 		prometheus.HistogramOpts{
 			Name:    "http_response_size_bytes",
 			Help:    "A histogram of response sizes for requests.",
 			Buckets: ResponseSizeBuckets,
 		},
-		[]string{},
+		[]string{"code", "method"},
+	)
+
+	// client metrics
+	clientCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "client_api_requests_total",
+			Help: "A counter for requests from the wrapped client.",
+		},
+		[]string{"client", "code", "method"},
+	)
+
+	clientDur = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "client_request_duration_seconds",
+			Help:    "A histogram of request latencies.",
+			Buckets: RequestDurationBucketsSeconds,
+		},
+		[]string{"client", "code", "method"},
+	)
+
+	clientInFlight = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "client_in_flight_requests",
+			Help: "A gauge of in-flight requests for the wrapped client.",
+		},
+		[]string{"client"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(
+		counter, duration, responseSize, // server metrics
+		clientCounter, clientDur, clientInFlight, // client metrics
 	)
+}
 
-	prometheus.MustRegister(counter, duration, responseSize)
+// NewGrpcServer returns a grpc server pre-configured with prometheus interceptors
+func NewGrpcServer() *grpc.Server {
```

Comment:
apologies for additional code movement here. `NewGrpcServer` is unchanged in this PR, just moved below the global vars.